### PR TITLE
Add a document for spike77 analysis about Node Name Uniqueness

### PR DESCRIPTION
**Pull request by emersonknapp**
Part of the solution for #77 (need to create implementation stories to close it)

Signed-off-by: Emerson Knapp <emerson.b.knapp@gmail.com>

### CODE REVIEW

File: spike-findings/77-node-name-uniqueness.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+Below are ideas in 2 categories:
+* mitigations: to make dealing with NUNNs less confusing
+* preventions: to avoid creating NUNNs in the first place
+
+The long-term status of NUNNs in ROS 2 is undefined, as seen in the Design Request #187, so preventions are more likely to get buy-in. However, reasonably simple mitigations are better than nothing given that the design has had no traction.
+
+### `ros2 node list` appended unique ID (mitigation)
+
+* Create a new `rmw` type `node_handle` that encapsulates a node name an arbitrary numerical unique identifier from the RMW implementation. Do not require RMW implementations to actually make this identifier unique, however for DDS implementations that already have a GUID, they can easily pass it out.
```
Comment (piraka9011) [inline spike-findings/77-node-name-uniqueness.md]: ...pass it on

File: spike-findings/77-node-name-uniqueness.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+### Services (mitigation)
+
+This is going to be very hard to fix uniquely. It would require creating unique topic name mappings for services, which affects the implementation of services top to bottom. More research would be required to determine the full scope of intervention required to do this, but it's unlikely that we'd even want to.
+
+Recommending punting this issue way out.
+
+### `ros2 launch` warn on node name overlap (prevention)
+
+ROS 2 "launch" at some point must have all Node actions expanded and their names fully substituted/qualified.
+At this point we should be able to scan the list for duplicate names and notify the user before launching them.
```
Comment (piraka9011) [inline spike-findings/77-node-name-uniqueness.md]: Do you know where/when this occurs in `launch`?

File: spike-findings/77-node-name-uniqueness.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+We could provide an extra CLI flag, either `--allow-duplicate-names` or `--error-on-duplicate-name`, depending on which behavior we want to be the default.
+
+The average mid-to-small ROS project is launched via a single top-level launchfile (often called `${PROJECT}_bringup.launch.py`), therefore this will catch many cases for the average user.
+Duplicate nodes that appear, if the launchfile did not warn them, must be from elsewhere on the network.
+This is useful information to know.
+
+NOTE for implementation:
+* `Node`s where the user does not explicitly set the name, will use the default name, which is not necessarily a duplicate case.
```
Comment (piraka9011) [inline spike-findings/77-node-name-uniqueness.md]: Can we enforce uniqueness for the default name at this layer?

File: spike-findings/77-node-name-uniqueness.md (hunk 1 above)
Comment (emersonknapp) [reply spike-findings/77-node-name-uniqueness.md]: Pass it through. Pass it forward, up, down, over.

I'm not sure I understand your question - the rmw implementation already contains a container of (DDS::GUID_t, node_name) pairs. 

What I'm specifying here is that they give us a representation of that GUID that we can print out. No specification of how they were created.

File: spike-findings/77-node-name-uniqueness.md (hunk 2 above)
Comment (emersonknapp) [reply spike-findings/77-node-name-uniqueness.md]: I do not

File: spike-findings/77-node-name-uniqueness.md (hunk 3 above)
Comment (emersonknapp) [reply spike-findings/77-node-name-uniqueness.md]: No, it's internal to the implementation of a given node, e.g. the C++ goes `rclcpp::Node("some_name")`

File: spike-findings/77-node-name-uniqueness.md (hunk 1 above)
Comment (piraka9011) [reply spike-findings/77-node-name-uniqueness.md]: My understanding from your bullet point is that the unique identifier will be the GUID from the DDS implementation. If this is the case, do we *still* need to worry about conflicts or is that handled by the DDS implementation?

I wasn't sure what you meant by pass it out 😅

Comment (emersonknapp) [reply spike-findings/77-node-name-uniqueness.md]: I still don't understand the question. Nodes will still have non-unique names - but with this solution we can just show them some little bit of text to visually differentiate them (to help improve confidence that the duplicate name is not an artifact of the cli tool)

Comment (emersonknapp) [reply spike-findings/77-node-name-uniqueness.md]: I've pushed another commit updating this section, can you let me know if it's more clear?

File: spike-findings/77-node-name-uniqueness.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+Things we should definitely do, in bang-for-buck order
+
+* Definitely: `ros2 node info` warn on querying NUNN
+* Definitely: `ros2 node list` note on listing NUNNs
+* Definitely: `ros2 launch` warnings when creating NUNNs
+
+Things we should consider doing, but need to start on quickly if they are to get in for Foxy
+
+* Consider: `ros2 node list` extra information
+* Consider: `ros2 node info` by unique ID
```
Comment (mm318) [inline spike-findings/77-node-name-uniqueness.md]: How about `ros2 node info` prints out info of all nodes that have the same name?

Comment (emersonknapp) [reply spike-findings/77-node-name-uniqueness.md]: Yes that's another way to solve that problem. We're not going to be able to do it for Foxy though so i don't think it matters, this isn't a design commitment document.

**Pull request merged**